File: golang-devildyw/middle/L46全排列.go
package main

func main() {}

func permute(nums []int) [][]int {
	result := make([][]int, 0)
	path := make([]int, 0)
	n := len(nums)

	set := make(map[int]bool)
	var dfs func()
	dfs = func() {
		if len(path) == n {
			tempResult := make([]int, len(path))
			copy(tempResult, path)
			result = append(result, tempResult)
			return
		}

		for i := 0; i < len(nums); i++ {
			if set[nums[i]] {
				continue
			}
			set[nums[i]] = true
			path = append(path, nums[i])
			dfs()
			path = path[:len(path)-1]
			set[nums[i]] = false
		}
	}

	dfs()
	return result
}

File: java-devildyw/src/main/java/top/dyw/leetcode/medium/L46全排列.java
package top.dyw.leetcode.medium;

import java.util.ArrayList;
import java.util.List;

public class L46全排列 {

    List<List<Integer>> result = new ArrayList();
    List<Integer> path = new ArrayList();
    boolean[] set;
    public List<List<Integer>> permute(int[] nums) {
        set = new boolean[nums.length];
        dfs(nums);

        return result;
    }

    public void dfs(int[]nums) {
        if (path.size() == nums.length) {
            result.add(new ArrayList(path));
            return;
        }

        for( int i = 0; i<nums.length; i++) {
            if (set[i]) {
                continue;
            }
            set[i] = true;
            path.add(nums[i]);
            dfs(nums);
            path.remove(path.size()-1);
            set[i] = false;
        }
    }

}
